[BE] Modernize `C10_UNUSED`

r-barnes opened the pull request:
[`[[maybe_unused]]`](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused) is part of C++17 standard 

Test Plan: Sandcastle





cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

albanD reviewed (approved):
Sounds fair as long as the oss macos, windows and arm builds are fine.

malfet reviewed (commented):
This looks fine in OSS, but I'm afraid some internal team could still be using ancient compiler versions (especially on Windows) and this will fail. But if it works, this is a super useful change

r-barnes committed:
[BE] Modernize `C10_UNUSED` (#138102)

Summary:
[`[[maybe_unused]]`](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused) is part of C++17 standard

Pull Request resolved: https://github.com/pytorch/pytorch/pull/138102

Test Plan:
Sandcastle

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

Reviewed By: malfet, HapeMask

Differential Revision: D64486813

Pulled By: r-barnes
